# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: toolkit/components/credentialmanagement/IdentityCredentialRequestManager.cpp
# Commit: 459739221a88
# Full Hash: 459739221a884b2b1c2adab42fd19fafe279bd20
# Author: Sandor Molnar <smolnar@mozilla.com>
# Date: 2025-06-05 04:07:37
# Regressor Bug: 1969017
# File Overlap Count: 2
# Description:
#   Revert "Bug 1969015, 1969017: apply code formatting via Lando" for causing build bustages
#   
#   This reverts commit ab681e36aeafc21ba62bb4680bdaa6206f02546d.
#   
#   Revert "Bug 1969017 - Refactor FedCM to use a managed actor rather than tagging onto WindowGlobalParent - r=anti-tracking-reviewers,emz"
# ==============================================================================

diff -r fc9986230203 -r 459739221a88 toolkit/components/credentialmanagement/IdentityCredentialRequestManager.cpp
--- a/toolkit/components/credentialmanagement/IdentityCredentialRequestManager.cpp	Thu Jun 05 00:15:19 2025 +0300
+++ b/toolkit/components/credentialmanagement/IdentityCredentialRequestManager.cpp	Thu Jun 05 00:19:41 2025 +0300
@@ -5,8 +5,11 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #include "IdentityCredentialRequestManager.h"
-#include "mozilla/ClearOnShutdown.h"
+#include "mozilla/dom/IdentityCredentialBinding.h"
+#include "mozilla/dom/WindowGlobalParent.h"
 #include "nsContentUtils.h"
+#include "nsNetUtil.h"
+#include "mozilla/BasePrincipal.h"
 
 namespace mozilla {
 
@@ -25,4 +28,68 @@
   return sSingleton;
 }
 
+nsresult IdentityCredentialRequestManager::StorePendingRequest(
+    const nsCOMPtr<nsIPrincipal>& aRPPrincipal,
+    const dom::IdentityCredentialRequestOptions& aRequest,
+    const RefPtr<
+        dom::IdentityCredential::GetIPCIdentityCredentialPromise::Private>&
+        aPromise,
+    const RefPtr<dom::CanonicalBrowsingContext>& aBrowsingContext) {
+  MOZ_ASSERT(aRPPrincipal);
+
+  for (const auto& provider : aRequest.mProviders) {
+    if (!provider.mLoginURL.WasPassed()) {
+      continue;
+    }
+    nsCOMPtr<nsIURI> idpOriginURI;
+    nsAutoCString idpOriginString;
+    if (provider.mOrigin.WasPassed()) {
+      idpOriginString = provider.mOrigin.Value();
+    } else {
+      // Infer the origin from the loginURL if one wasn't provided
+      idpOriginString = provider.mLoginURL.Value();
+    }
+    nsresult rv = NS_NewURI(getter_AddRefs(idpOriginURI), idpOriginString);
+    if (NS_WARN_IF(NS_FAILED(rv))) {
+      return NS_ERROR_DOM_BAD_URI;
+    }
+    nsCOMPtr<nsIPrincipal> idpPrincipal = BasePrincipal::CreateContentPrincipal(
+        idpOriginURI, aRPPrincipal->OriginAttributesRef());
+
+    NS_ENSURE_TRUE(idpPrincipal, NS_ERROR_FAILURE);
+    nsTArray<PendingRequestEntry>& list =
+        mPendingRequests.LookupOrInsert(idpPrincipal);
+    list.AppendElement(PendingRequestEntry(aRPPrincipal, aRequest, aPromise,
+                                           aBrowsingContext));
+  }
+  return NS_OK;
+}
+
+void IdentityCredentialRequestManager::NotifyOfStoredCredential(
+    const nsCOMPtr<nsIPrincipal>& aIDPPrincipal,
+    const dom::IPCIdentityCredential& aCredential) {
+  MOZ_ASSERT(aIDPPrincipal);
+  auto listLookup = mPendingRequests.Lookup(aIDPPrincipal);
+  if (listLookup) {
+    for (auto& entry : listLookup.Data()) {
+      if (!entry.mBrowsingContext) {
+        continue;
+      }
+      // We must (asynchronously) test if this credential should be sent down
+      // to the site.
+      dom::IdentityCredential::AllowedToCollectCredential(
+          entry.mRPPrincipal, entry.mBrowsingContext, entry.mRequestOptions,
+          aCredential)
+          ->Then(
+              GetCurrentSerialEventTarget(), __func__,
+              [aCredential, entry](bool effectiveCredential) {
+                if (effectiveCredential) {
+                  entry.mPromise->Resolve(aCredential, __func__);
+                }
+              },
+              []() {});
+    }
+  }
+}
+
 }  // namespace mozilla